File: 1524-number-of-sub-arrays-with-odd-sum/1524-number-of-sub-arrays-with-odd-sum.cpp
class Solution {
public:
    int numOfSubarrays(vector<int>& arr) {
        long total = 0;
        int hmap[2];
        // keep track of even prefixSum
        hmap[0] = 1;
        // keep track of odd prefixSum
        hmap[1] = 0;
        int pSum = 0;
        for( int i = 0; i < arr.size(); i++ ) {
            pSum += arr[i];
            
            if( pSum % 2 == 0 ) 
                total += hmap[1];
            else
                total += hmap[0];
            
            // update the counts
            if( pSum % 2 == 0 ) 
                hmap[0] += 1;
            else
                hmap[1] += 1;
            
            total = total % 1000000007;
        }
        
        return total;    
    }
};

File: 1524-number-of-sub-arrays-with-odd-sum/1524-number-of-sub-arrays-with-odd-sum.py
class Solution(object):
    def numOfSubarrays(self, arr):
        """
        :type arr: List[int]
        :rtype: int
        """
        total = 0
        hmap = {}
        hmap["odd"] = 0
        hmap["even"] = 1
        pSum = 0
        
        for i in range(len(arr)):
            pSum += arr[i]
            
            if pSum % 2 == 0:
                total += hmap["odd"] 
            else:
                total += hmap["even"]
            
            total = total % 1000000007 
            
            if pSum % 2 == 0:
                hmap["even"] += 1
            else:
                hmap["odd"] += 1
        
        return total
            
            
            
            